### PR TITLE
Feat/site launch

### PR DESCRIPTION
**Pull request by kishore03109**
## Problem

Automate site launch process

Partially Addresses #516 

Steps to test locally: 

1. Run database migrations by running `npx sequelize-cli db:migrate`
2. Populate the `sites` table database with at least one site
3. Run through the steps [here](https://github.com/isomerpages/isomercms-backend/tree/feat/add-serverless#running-locally) to start local stack and step functions 
4. Create incoming and outgoing queues by running: `aws --endpoint-url=<URL_OF_LOCALSTACK> sqs create-queue --queue-name incomingQueue` and `aws --endpoint-url=<URL_OF_LOCALSTACK> sqs create-queue --queue-name outgoingQueue`
5. Update `INCOMING_QUEUE_URL` and `OUTGOING_QUEUE_URL` in .env file
6. Start backend server 
7. Submit request to launch site via FormSG
8. Check to see if the database tables have been updated
9. Check Amplify console to see domain being added 

## Deploy Notes

This PR involves addition of two tables in database. This would require DB migrations when merging to prod/staging. As discussed in the eng sync, there was a presence of a indeterministic bug (similar issue documented [here](https://github.com/aws/aws-sdk-js/issues/1453) and [here](https://github.com/aws/aws-sdk-js/issues/2793)) which I have tried to fix this in the commit [here](https://github.com/isomerpages/isomercms-backend/pull/517/commits/e9f3e06471119c35c7e2ddf18a24cd2ea86b2e17), but the issue is still not fixed :( . This is slightly scary as wrapping the error exists outside of try/catch block, so in the worst case, this will crash the server. Another dirty solution to be to modify the aws sdk as described [here](https://github.com/aws/aws-sdk-js/issues/2700). Running this on staging for a prolonged period of time (maybe a day/over the weekend) with some manual integration tests might be important before merging to prod. ._. 

**New environment variables**:

- `INCOMING_QUEUE_URL` : incoming queue's URL 
- `OUTGOING_QUEUE_URL` : outgoing queue's URL 

TODO before merging to prod: 
[ ] Remove `isDev` flag [here](https://github.com/isomerpages/isomercms-backend/blob/feat/site-launch/src/routes/formsgSiteLaunch.ts#L67) 
[ ] Fix network issue in `Request.HTTP_DATA` in aws SDK

### CODE REVIEW

File: src/services/infra/InfraService.ts [hunk 1]
```diff
@@ -91,4 +104,123 @@ export default class InfraService {
...
+        `Created Domain association for ${repoName} to ${primaryDomain}`
+      )
+
+      // Get DNS records from Amplify
+      const dnsInfo = await this.launchesService.getDomainAssociationRecord(
+        primaryDomain,
+        appId
+      )
+
+      const certificationRecord = dnsInfo.domainAssociation?.certificateVerificationDNSRecord?.split(
```
Comment (dcshzj) [inline src/services/infra/InfraService.ts:137]: If `dnsInfo.domainAssociation?.certificateVerificationDNSRecord` is null, would that cause the `split()` to fail?

Comment (dcshzj) [inline src/services/infra/InfraService.ts:137]: Is it possible to add a comment to give an example of how the `certificationRecord` should look like?

File: src/services/infra/InfraService.ts [hunk 2]
```diff
@@ -91,4 +104,123 @@ export default class InfraService {
...
+        " "
+      )
+
+      if (!certificationRecord) {
+        throw Error("certification record not created yet")
+      }
+
+      const subDomainList = dnsInfo.domainAssociation?.subDomains
```
Comment (dcshzj) [inline src/services/infra/InfraService.ts:145]: Same for this, would be nice to see how the `subDomainList` looks like.

**Review by dcshzj — CHANGES_REQUESTED** [review]

File: src/database/migrations/20220920084207-create-redirects.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      launch_id: {
+        allowNull: false,
+        type: Sequelize.INTEGER,
+      },
+      type: {
+        allowNull: false,
+        type: Sequelize.STRING,
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/database/migrations/20220920084207-create-redirects.js]: [fd65a8d](https://github.com/isomerpages/isomercms-backend/pull/517/commits/fd65a8d05440a086e41ffdf76f8502bdf6700af0)

File: src/routes/formsgSiteLaunch.ts [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    Record<string, never>,
+    { data: { submissionId: string } },
+    never,
+    { submission: DecryptedContent }
+  > = async (req, res) => {
+    // 1. Extract arguments
+    const { submissionId } = req.body.data
+    const { responses } = res.locals.submission
+
+    // todo change back these variables to const
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/routes/formsgSiteLaunch.ts:52]: yes opps this faeture is still in development - will remove when merging to prod ._.

File: src/services/infra/InfraService.ts (hunk 2 above)
Comment (kishore03109) [reply src/services/infra/InfraService.ts:161]: [be86a41](https://github.com/isomerpages/isomercms-backend/pull/517/commits/be86a41d7454613f57462ce63876b32150f86b0e)

File: src/services/infra/InfraService.ts (hunk 1 above)
Comment (kishore03109) [reply src/services/infra/InfraService.ts]: [be86a41](https://github.com/isomerpages/isomercms-backend/pull/517/commits/be86a41d7454613f57462ce63876b32150f86b0e)

File: src/services/identity/QueueClient.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+import logger from "@root/logger/logger"
+
+const { INCOMING_QUEUE_URL, OUTGOING_QUEUE_URL } = process.env
+export default class QueueClient {
+  private readonly sqs: AWS.SQS
+
+  private readonly incomingQueueUrl
+
+  private readonly outgoingQueueUrl
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/identity/QueueClient.ts]: [fc399af](https://github.com/isomerpages/isomercms-backend/pull/517/commits/fc399af2a49269a58737609c8c5477befa996740)

File: src/services/identity/LaunchesService.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+  private readonly launchClient: LaunchClient
+
+  constructor({
+    deployment,
+    launches,
+    repo,
+    redirections,
+  }: LaunchesServiceProps) {
+    this.deployment = deployment
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/identity/LaunchesService.ts]: [fc399af](https://github.com/isomerpages/isomercms-backend/pull/517/commits/fc399af2a49269a58737609c8c5477befa996740)

File: src/services/identity/LaunchesService.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+import { ModelStatic } from "sequelize"
+
+import logger from "@logger/logger"
+
+import { Deployment, Launches, Repo, Site, User } from "@database/models"
+import { Redirections } from "@root/database/models/Redirections"
+import { AmplifyError } from "@root/types/index"
+import LaunchClient from "@services/identity/LaunchClient"
+
+type launchesCreateParamsType = Partial<Launches> & {
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/identity/LaunchesService.ts:12]: [fc399af](https://github.com/isomerpages/isomercms-backend/pull/517/commits/fc399af2a49269a58737609c8c5477befa996740)

File: src/routes/formsgSiteLaunch.ts [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    if (isDev) {
+      requesterEmail = "kishore@open.gov.sg"
+      repoName = "kishore-test"
+      primaryDomain = "kishoretest.isomer.gov.sg"
+      redirectionDomain = "www.kishoretest.isomer.gov.sg"
+      agencyEmail = "kishore@open.gov.sg"
+    }
+
+    logger.info(
+      `Lauch site form submission [${submissionId}] (repoName '${repoName}', domain '${primaryDomain}') requested by <${requesterEmail}>`
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/routes/formsgSiteLaunch.ts]: [fc399af](https://github.com/isomerpages/isomercms-backend/pull/517/commits/fc399af2a49269a58737609c8c5477befa996740)

File: src/services/identity/LaunchClient.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  createGetDomainAssociationCommandInput = (
+    appId: string,
+    domainName: string
+  ): GetDomainAssociationCommandInput => ({
+    appId,
+    domainName,
+  })
+
+  sendGetDomainAssociationCommandInput = (
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/identity/LaunchClient.ts]: [fc399af](https://github.com/isomerpages/isomercms-backend/pull/517/commits/fc399af2a49269a58737609c8c5477befa996740)

File: src/services/identity/LaunchesService.ts [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+     * todo: add some level of retry logic if get domain association command
+     * does not contain the DNS redirections info.
+     */
+    return this.launchClient.sendGetDomainAssociationCommandInput(
+      getDomainAssociationOptions
+    )
+  }
+
+  async updateLaunchTable(
+    updateParams: Partial<Launches> & { id: Site["id"] }
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/identity/LaunchesService.ts]: [fd65a8d](https://github.com/isomerpages/isomercms-backend/pull/517/commits/fd65a8d05440a086e41ffdf76f8502bdf6700af0)

File: src/services/infra/InfraService.ts [hunk 11]
```diff
@@ -20,14 +26,21 @@ export default class InfraService {
...
     sitesService,
     reposService,
     deploymentsService,
+    launchesService,
   }: InfraServiceProps) {
     this.sitesService = sitesService
     this.reposService = reposService
     this.deploymentsService = deploymentsService
+    this.launchesService = launchesService
+    this.queueService = new QueueService()
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/infra/InfraService.ts:43]: [7f61a14](https://github.com/isomerpages/isomercms-backend/pull/517/commits/7f61a143610c846cb9cc7fdfafcdc4967eda73c1)

File: src/services/identity/LaunchesService.ts [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+  private readonly launchClient: LaunchClient
+
+  constructor({
+    deployment,
+    launches,
+    repo,
+    redirections,
+  }: LaunchesServiceProps) {
+    this.deployment = deployment
+    this.launchClient = new LaunchClient()
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/identity/LaunchesService.ts]: [7f61a14](https://github.com/isomerpages/isomercms-backend/pull/517/commits/7f61a143610c846cb9cc7fdfafcdc4967eda73c1)

**Pull request merged**